Generator with await not considered an AsyncGenerator

<!--
  If you're new to mypy and you're not sure whether what you're experiencing is a mypy bug, please see the "Question and Help" form
  instead.
-->

**Bug Report**

<!--
Note: If the problem you are reporting is about a specific library function, then the typeshed tracker is better suited
for this report: https://github.com/python/typeshed/issues
-->

- Generators containing the `await` keyword are considered Generator, rather than AsyncGenerator
- Generator syntax does not permit `await` in a non-async function

**To Reproduce**

test.py:
```python
import asyncio
from typing import AsyncGenerator


async def get_value(v: int) -> int:
    await asyncio.sleep(1)
    return v + 1


def get_generator() -> AsyncGenerator[int, None]:
    return (await get_value(v) for v in [1, 2, 3])


async def test() -> None:
    print(type(get_generator()))
    async for x in get_generator():
        print(x)


asyncio.get_event_loop().run_until_complete(test())
```
python:
`> py test.py`
```
<class 'async_generator'>
2
3
4
```
mypy:
`mypy test.py`
```
test.py:11: error: Incompatible return value type (got "Generator[int, None, None]", expected "AsyncGenerator[int, None]")
```

**Expected Behavior**

No errors should be reported.

**Actual Behavior**

```
test.py:11: error: Incompatible return value type (got "Generator[int, None, None]", expected "AsyncGenerator[int, None]")
test.py:11: error: 'await' outside coroutine ('async def')
```

**Your Environment**

<!-- Include as many relevant details about the environment you experienced the bug in -->

- Mypy version used: 0.812
- Mypy command-line flags: None
- Mypy configuration options from `mypy.ini` (and other config files):
```

[mypy]
plugins = pydantic.mypy

strict_optional = True
warn_redundant_casts = True
warn_unused_ignores = True
disallow_any_generics = True
check_untyped_defs = True
disallow_untyped_defs = True

[pydantic-mypy]
init_forbid_extra = True
warn_required_dynamic_aliases = True
warn_untyped_fields = True

[mypy-bson.*]
ignore_missing_imports = True

[mypy-cachetools.*]
ignore_missing_imports = True

[mypy-asyncache.*]
ignore_missing_imports = True

[mypy-loguru]
ignore_missing_imports = True

[mypy-aiohttp_xmlrpc.*]
ignore_missing_imports = True

[mypy-asyncpg.*]
ignore_missing_imports = True

[mypy-bcrypt]
ignore_missing_imports = True

[mypy-passlib.*]
ignore_missing_imports = True

[mypy-slugify.*]
ignore_missing_imports = True

[mypy-barcode.*]
ignore_missing_imports = True

[mypy-qrcode.*]
ignore_missing_imports = True

[mypy-furl]
ignore_missing_imports = True
although 
[mypy-defusedxml.*]
ignore_missing_imports = True

[mypy-pyticketswitch.*]
ignore_missing_imports = True

[mypy-pymongo.*]
ignore_missing_imports = True

[mypy-motor.*]
ignore_missing_imports = True

[mypy-stackprinter]
ignore_missing_imports = True

[mypy-PIL.*]
ignore_missing_imports = True

[mypy-aztec_code_generator.*]
ignore_missing_imports = True

[mypy-pdf417.*]
ignore_missing_imports = True

[mypy-pylibdmtx.*]
ignore_missing_imports = True


[mypy-toolz]
ignore_missing_imports = True

[mypy-stripe]
ignore_missing_imports = True

[mypy-google.*]
ignore_missing_imports = True

[mypy-coinoxr]
ignore_missing_imports = True

[mypy-sendgrid.*]
ignore_missing_imports = True

[mypy-opencensus.*]
ignore_missing_imports = True
```
- Python version used: 3.9.2, acquired via pyenv
- Operating system and version: Ubuntu 20.04.2

<!--
You can freely edit this text, please remove all the lines
you believe are unnecessary.
-->

Simple repro with mypy latest: https://mypy-play.net/?mypy=master&python=3.12&gist=71cdbfcc61179de3290480733083180c&flags=warn-unreachable

```python
async def something(i: int) -> tuple[None, ...]:
    return (None,)

async def main() -> None:
    async_or_nah = (
        thing
        for i in range(5)
        for thing in await something(i)
    )
    
    
    async for blah in async_or_nah: # error: "Generator[None, None, None]" has no attribute "__aiter__" (not async iterable)  [attr-defined]
        print(blah)
```

--- a/mypy/checkexpr.py
+++ b/mypy/checkexpr.py
@@ -5585,8 +5585,13 @@ class ExpressionChecker(ExpressionVisitor[Type]):
 
     def visit_generator_expr(self, e: GeneratorExpr) -> Type:
         # If any of the comprehensions use async for, the expression will return an async generator
-        # object, or if the left-side expression uses await.
-        if any(e.is_async) or has_await_expression(e.left_expr):
+        # object, or await is used anywhere but in the leftmost sequence.
+        if (
+            any(e.is_async)
+            or has_await_expression(e.left_expr)
+            or any(has_await_expression(sequence) for sequence in e.sequences[1:])
+            or any(has_await_expression(cond) for condlist in e.condlists for cond in condlist)
+        ):
             typ = "typing.AsyncGenerator"
             # received type is always None in async generator expressions
             additional_args: list[Type] = [NoneType()]

--- a/test-data/unit/check-async-await.test
+++ b/test-data/unit/check-async-await.test
@@ -573,6 +573,25 @@ async def return_f() -> AsyncGenerator[int, None]:
 [builtins fixtures/dict.pyi]
 [typing fixtures/typing-async.pyi]
 
+[case testImplicitAsyncGenerator]
+from typing import List
+
+async def get_list() -> List[int]:
+    return [1]
+
+async def predicate() -> bool:
+    return True
+
+async def test_implicit_generators() -> None:
+    reveal_type(await predicate() for _ in [1])  # N: Revealed type is "typing.AsyncGenerator[builtins.bool, None]"
+    reveal_type(x for x in [1] if await predicate())  # N: Revealed type is "typing.AsyncGenerator[builtins.int, None]"
+    reveal_type(x for x in await get_list())  # N: Revealed type is "typing.Generator[builtins.int, None, None]"
+    reveal_type(x for _ in [1] for x in await get_list())  # N: Revealed type is "typing.AsyncGenerator[builtins.int, None]"
+
+[builtins fixtures/dict.pyi]
+[typing fixtures/typing-async.pyi]
+
+
 -- The full matrix of coroutine compatibility
 -- ------------------------------------------
 

